[gzip]: allow gzip to work w/ http backend w/o content-lenght

tehnerd opened the pull request:
Signed-off-by: Nikita V. Shirokov <tehnerd@tehnerd.com>


For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message:
this is an attempt to address issue in #14121
tl;dr is that some http/2 backends does not send "content-length" header in replies, as http/2 spec do have the same info as a part of the frame (unfortunately it seems like there is no way to pass this value from the frame to the filter) and transfer-encoding=chunked (before this diff having that header/encoding was a prerequisite, if content-length is not defined) was removed from http/2 spec.
As discussed in the issue itself, instead, if there is no content lengths header - we would try to gzip it by default.
this new behavior is controlled by runtime guarded feature envoy.reloadable_features.enable_compression_without_chunked_header

Additional Description:
This is rework of PR#14405. Compare to original PR this one has proper handling of websockets (or any other) upgrade requests.
Compare to original PR this one includes:
1. additional is(H2)Upgrade checks in encode/decodeHeaders functions in compressor filter
2. websockets integration tests now contains compression filter config

Risk Level:
Low
Testing:
same tests as in original PR#14405 plus 
websocket's integration tests now contains compression filter and passing (w/o isUpgrade fix and w/ compression filter this ITs were failing with:
```
 Actual: 0x7fff5ae54678 (of type Envoy::Http::TestResponseHeaderMapImpl*), 
==============================Expected header map:==============================
':status', '101'
'connection', 'upgrade'
'upgrade', 'websocket'
=======================is not equal to actual header map:=======================
':status', '101'
'connection', 'upgrade'
'upgrade', 'websocket'
'vary', 'Accept-Encoding'
================================================================================
```

Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

tehnerd commented:
cc @rgs1 @alyssawilk

rojkov commented on test/integration/websocket_integration_test.cc:
```diff
@@ -133,6 +133,15 @@ void WebsocketIntegrationTest::initialize() {
+    name: test
+    typed_config:
+      "@type": type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip
+)EOF");
```
Would it make sense to add a similar change to the http2 integration test? To make sure compression actually works for h2?

alyssawilk replied:
While my intuition is it shouldn't be done, I don't see anything in either spec about legality of compressing HTTP/1 connect payload which makes me wonder if it's legal, and if it is legal if we should support it by default.

https://tools.ietf.org/html/rfc7540#section-8.3
https://tools.ietf.org/html/rfc8441

@PiotrSikora @lizan thoughts?

alyssawilk replied:
cc @DavidSchinazi

alyssawilk replied:
actually @rgs1 since you reported the issue in the first place, do you know if it broke WS on HTTP/1, HTTP/2 or both?  If it's not supported globally it'd be a useful data point.

rgs1 replied:
on HTTP/1 (downstream & upstream)

alyssawilk replied:
Fair.  Either way I agree that integration tests for HTTP/2 would be good to have.

DavidSchinazi replied:
Transfer-Encoding is not allowed on CONNECT responses. [RFC 7230 s3.3.1](https://tools.ietf.org/html/rfc7230#section-3.3.1) states:
> A server MUST NOT send a Transfer-Encoding header field in any 2xx (Successful) response to a CONNECT request

tehnerd replied:
Sorry. i little bit lost here: are we talking about H/2 integration test for WS or ? If former - i think initializer (where this new compression config is added) is protocol agnostic and most likely used in all of the integration tests (and i assume we do have h/2 ws)

alyssawilk replied:
Yeah, if it runs for HTTP/1 and HTTP/2 you're fine here.
Unfortunately I think including extensions in core tests is legal so I think you're going to have to add similar test to the compressor lib.

tehnerd replied:
actually looks like we do not have h2 integration tests for websockets. Looks like you were about to add them in cd171d9a95f551628453c83f0b00924e8b310bf0 (at least this is where WebsocketTestParams struct were introduced) but it is actually not being used anywhere so all h2 ws upgrade path is untested right now

alyssawilk commented on docs/root/version_history/current.rst:
```diff
@@ -64,6 +64,7 @@ Removed Config or Runtime
 
 New Features
 ------------
+* compression: added new runtime guard `envoy.reloadable_features.enable_compression_without_chunked_header` (enabled by default) to control compression if no transfer-encoding=chunked header exists (now envoy would compress in such cases) to account for http/2 connections, where such transfer-encoding were removed.
```
compression: extended the compression allow compressing when the content length header is not present.  This behavior may be temporarily reverted by setting `envoy.reloadable_features.enable_compression_without_chunked_header` to false.

alyssawilk commented on source/extensions/filters/http/common/compressor/compressor.cc:
```diff
@@ -495,6 +497,10 @@ bool CompressorFilter::isEtagAllowed(Http::ResponseHeaderMap& headers) const {
 
 bool CompressorFilterConfig::DirectionConfig::isMinimumContentLength(
     const Http::RequestOrResponseHeaderMap& headers) const {
+  if (StringUtil::caseFindToken(headers.getTransferEncodingValue(), ",",
```
can this ever be true? I think the transfer encoding headers are removed before the filter ever sees it. It would also result in HTTP/1.1 and HTTP/2 responses without content length being treated differently.  I think this was buggy prior to your change but I'd be inclined to fix and test while we're in here.

tehnerd replied:
why h1 and h2 would be different? (i this diff we do not distinct h1 vs h2 in compressor. the runtime feature just either permit or deny compression if no content-length header). and yes it seems that you are right. at least manual testing shows that there is no compression if only trunsfer-encoding=chunked present. do you propose by somehow passing transfer-encoding into the filter or just remove this condition? I think originally idea was that is it is chunked - we making assumption that response is big enough and worth compressing (the whole idea behind isMinimumContentLength (from my understanding) is to be able to do some kind of performance optimization (do not compress something which is really small)) so from my point of view we can just omit this logic (unless @rojkov has some other thoughts on this)

alyssawilk replied:
So HTTP/1 on the wire, has a transfer-encoding: chunked header for (most) responses without a content length.  The alternate is framing by connection close.
HTTP/2 disallows the transfer encoding header though data is implicitly chunked up by HTTP/2 data frames.

In Envoy, the HCM removes any transfer encoding header before this code is called.  If you add an integration test with T-E:chunked, and an assert here, I believe it can never trigger.
I believe the code needs to simply have one behavior if content length is present, and one if content length is absent, and not rely on the transfer encoding header.

rojkov replied:
This code is quite old. I guess the original thinking was indeed to assume the payload is big enough if `transfer-encoding: chunked` is present.

> In Envoy, the HCM removes any transfer encoding header before this code is called.

Is it about the `chunked` part only or fully removed? If any t-e header is removed in HCM then also `CompressorFilter::isTransferEncodingAllowed()` becomes useless.

alyssawilk replied:
Hm, I thought it was removed early but looks like I may be incorrect:

https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_utility.cc#L398
response_headers.removeTransferEncoding();

called here:
https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_impl.cc#L1385

which looks like it's right before it's sent the codec.  Either way a test would be good to make sure we're consistent between HTTP/1 and HTTP/2 without content length, SG?

tehnerd replied:
so my understanding here is that before any filters in http conn manager we are calling this: https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_impl.cc#L1362
and especially this: https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_impl.cc#L1385

which seems like a helper which always remove TE headers: https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_utility.cc#L398

so it feels that it removes all TE. not just chunked. however i could miss some corner case as i'm not super familiar with conn_manager code path

tehnerd replied:
What tests are we talking about? for the WS/Update (to make sure that h1 and h2 are the same) or something else?

alyssawilk commented on source/extensions/filters/http/common/compressor/compressor.cc:
```diff
@@ -221,10 +223,10 @@ Http::FilterHeadersStatus CompressorFilter::encodeHeaders(Http::ResponseHeaderMa
-                              !hasCacheControlNoTransform(headers) && isEtagAllowed(headers) &&
-                              !headers.getInline(response_content_encoding_handle.handle());
+  const bool isCompressible =
+      isEnabledAndContentLengthBigEnough && !Http::Utility::isUpgrade(headers) &&
```
shouldn't the new upgrade also be runtime guarded?  maybe have a temporary bool for checking this?

tehnerd replied:
sure will add runtime check here as well

alyssawilk commented on source/extensions/compression/gzip/compressor/BUILD:
```diff
@@ -26,6 +26,9 @@ envoy_cc_extension(
     name = "config",
     srcs = ["config.cc"],
     hdrs = ["config.h"],
+    extra_visibility = [
```
Unfortunately the reason this wasn't visible is we don't allow Envoy core code to depend on extensions.

tehnerd replied:
just for the reference (mostly for myself so it would be easier to find in future). this should be reworked because of https://github.com/envoyproxy/envoy/issues/9953

alyssawilk replied:
Yeah, it's a relatively new policy.  Before we occasionally broke people's builds if they opted out of certain extensions, then fixed it after they called it out.  Better to not break in the first place :-P

If there's a place we could call that out where you would have seen it, please improve the docs in this area.  It's totally intuitive if something fails to build due to visibility to just fix it.  Normally we put such checks in the format scripts, but unfortunately we can't with this one until we clean up all the legacy visibility rules.

tehnerd replied:
changes visibility to test/extensions. i hope this is ok (for test/extensions to depends on source/extensions)?

alyssawilk replied:
why is it needed? I think we also don't want cross-extension dependencies, so explicit lists would be better.  If that's onerous we could at least say //test/extensions/filters/http:__subpackages__ ? I'll defer to @lizan if he has thoughts/preferences.

tehnerd replied:
yeah. looks like i still can use this target in tests even w/o explicit "extra_visibility". will remove

alyssawilk reviewed (commented):
Thanks for jumping on this fix so fast!

tehnerd commented on test/extensions/filters/http/common/compressor/websocket_with_compressor_integration_test.cc:
```diff
@@ -0,0 +1,248 @@
+#include "test/extensions/filters/http/common/compressor/websocket_with_compressor_integration_test.h"
```
this (and .h) is copy/pasted test/integration/websocket_integration_test (i've just removed few things which are not related to upgrade and renamed class/structs by adding WithCompressor (as well as added compressor configuration itself))

tehnerd commented:
@alyssawilk @rojkov somehow lost in translation/discussion. What does it take for this to move forward? there were some discussion about itest, but i missed the part - were we talking about WS related tests (which is done) or there were some other requirments (e.g. IT for transfer-encoding etc)?

alyssawilk commented:
I think I mainly want an integration test that WS doesn't get zipped for
HTTP/1 or HTTP/2, and we do a manual check that the prior PR would fail
that test, to make sure we're regression testing the old bug.

If we have that test, I think you can just do a master merge and make sure
CI is happy, and I'm up for doing another pass!

On Thu, Jan 14, 2021 at 12:34 PM Nikita V. Shirokov <
notifications@github.com> wrote:

> @alyssawilk <https://github.com/alyssawilk> @rojkov
> <https://github.com/rojkov> somehow lost in translation/discussion. What
> does it take for this to move forward? there were some discussion about
> itest, but i missed the part - were we talking about WS related tests
> (which is done) or there were some other requirments (e.g. IT for
> transfer-encoding etc)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/14584#issuecomment-760348450>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AELALPP6IYZPXIWMYCSPCBTSZ4TLNANCNFSM4VYC4IDQ>
> .
>

alyssawilk commented:
Oh, sorry, also making sure we integration test that both HTTP/1 and HTTP2
handle compressing when content length is absent (i.e. rely on the absence
of the C-L header, rather than the presence of the T-E header)

On Thu, Jan 14, 2021 at 2:53 PM Alyssa (Rzeszutek) Wilk <alyssar@google.com>
wrote:

> I think I mainly want an integration test that WS doesn't get zipped for
> HTTP/1 or HTTP/2, and we do a manual check that the prior PR would fail
> that test, to make sure we're regression testing the old bug.
>
> If we have that test, I think you can just do a master merge and make sure
> CI is happy, and I'm up for doing another pass!
>
> On Thu, Jan 14, 2021 at 12:34 PM Nikita V. Shirokov <
> notifications@github.com> wrote:
>
>> @alyssawilk <https://github.com/alyssawilk> @rojkov
>> <https://github.com/rojkov> somehow lost in translation/discussion. What
>> does it take for this to move forward? there were some discussion about
>> itest, but i missed the part - were we talking about WS related tests
>> (which is done) or there were some other requirments (e.g. IT for
>> transfer-encoding etc)?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/envoyproxy/envoy/pull/14584#issuecomment-760348450>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AELALPP6IYZPXIWMYCSPCBTSZ4TLNANCNFSM4VYC4IDQ>
>> .
>>
>

tehnerd commented:
> I think I mainly want an integration test that WS doesn't get zipped for HTTP/1 or HTTP/2, and we do a manual check that the prior PR would fail that test, to make sure we're regression testing the old bug. If we have that test, I think you can just do a master merge and make sure CI is happy, and I'm up for doing another pass!
> […](#)

This one is done. As i've mentioned in "Testing" part of PR w/o new checks WS integration test fails as it starts to have "Accept-Encoding: vary" header added by compression filter

tehnerd commented:
> Oh, sorry, also making sure we integration test that both HTTP/1 and HTTP2 handle compressing when content length is absent (i.e. rely on the absence of the C-L header, rather than the presence of the T-E header) On Thu, Jan 14, 2021 at 2:53 PM Alyssa (Rzeszutek) Wilk <alyssar@google.com> wrote:
> […](#)

T-E:chunked logic was removed from compression filter. so current logic is 

1. check if c-l present. if so - check if it is bigger than configured minimal value
2. if not - compress by default.

and another check which is making sure that request is not "update"  (by reusing Utilit::isUpgrade).

so current version of the filter is http version agnostic (as t-e was removed)




going to do merge with muster and update the diff

tehnerd commented on docs/root/version_history/current.rst:
```diff
@@ -34,6 +34,45 @@ Removed Config or Runtime
 
 New Features
 ------------
+* compression: extended the compression allow compressing when the content length header is not present. This behavior may be temporarily reverted by setting `envoy.reloadable_features.enable_compression_without_content_length_header` to false.
```
hmm looks like bad automerge. fill fix in a few

rojkov commented on source/extensions/filters/http/common/compressor/compressor.cc:
```diff
@@ -506,9 +521,12 @@ bool CompressorFilterConfig::DirectionConfig::isMinimumContentLength(
   }
-
-  return StringUtil::caseFindToken(headers.getTransferEncodingValue(), ",",
-                                   Http::Headers::get().TransferEncodingValues.Chunked);
```
nit, feel free to ignore: I don't have strong opinion, but it wouldn't harm to just move this check above the C-L check as it was in the previous PR. T-E is not removed from response headers by the time this function is called. My only concern is that disabling this feature flag for whatever reason doesn't bring us to the old behavior.

tehnerd replied:
it was removed because filter does not see this header (T-E:chunked). there is above discussion w/ @alyssawilk  about it and that i've tried to run a test w/ ASSERT in this code path (basically if (StringUtil::case...){ ASSERT(1 == 2)}) generated a reply w/ T-E:chunked and it was:

rojkov replied:
In the discussion @alyssawilk confirmed that the filter should see the header actually and in my tests the filter sees the header very well. Usually to play with response headers I use this fake HTTP server:
```python
#!/usr/bin/env python3
"""
Very simple HTTP server in python for logging requests
Usage::
    ./server.py [<port>]
"""
from http.server import BaseHTTPRequestHandler, HTTPServer
import logging
import gzip
import shutil


class S(BaseHTTPRequestHandler):

  def _set_response(self):
    self.send_response(200)
    self.send_header('Content-Type', 'text/html')
    self.end_headers()

  def _get_chunk_size(self):
      size_str = self.rfile.read(2)
      while size_str[-2:] != b"\r\n":
          size_str += self.rfile.read(1)
      return int(size_str[:-2], 16)

  def _get_chunk_data(self, chunk_size):
      data = self.rfile.read(chunk_size)
      self.rfile.read(2) # read out '\r\n'
      return data

  def do_GET(self):
    logging.info("GET request,\nPath: %s\nHeaders:\n%s\n", str(self.path), str(self.headers))
    msg = "GET request for {}\nLorem ipsum is a pseudo-Latin text used in web design, typography,\n layout, and printing in place of English\n to emphasise design elements over content.\n It's also called placeholder (or filler)\n text. It's a convenient tool for mock-ups\n".format(
        self.path).encode('utf-8')
    self.send_response(200)
    self.send_header('Content-Type', 'text/plain')
    #self.send_header('Content-Length', str(len(msg)))
    self.send_header('Transfer-Encoding', 'chunked')
    #self.send_header('vary', 'Accept-Encoding')
    self.end_headers()
    self.wfile.write(msg)

  def do_POST(self):
      logging.info("POST request,\nPath: %s\nHeaders:\n%s", str(self.path), str(self.headers))
      if self.headers['Content-Length'] is not None and \
         self.headers['Content-Encoding'] is None:
          content_length = int(self.headers['Content-Length'])  # <--- Gets the size of data
          post_data = self.rfile.read(content_length)  # <--- Gets the data itself
          logging.info("Body:\n%s\n", post_data.decode('utf-8'))

      if self.headers['Content-Encoding'] == 'gzip':
          data = b""
          while True:
              chunk_size = self._get_chunk_size()
              if chunk_size == 0:
                  break
              else:
                  data += self._get_chunk_data(chunk_size)
          logging.info("Uncompressed Body:\n%s\n", gzip.decompress(data).decode('utf-8'))

      self.send_response(200)
      self.send_header('Content-Type', 'text/html')
      self.end_headers()
      self.wfile.write("POST request for {}\n".format(self.path).encode('utf-8'))


def run(server_class=HTTPServer, handler_class=S, port=8080):
  logging.basicConfig(level=logging.INFO)
  server_address = ('', port)
  httpd = server_class(server_address, handler_class)
  logging.info('Starting httpd...\n')
  try:
    httpd.serve_forever()
  except KeyboardInterrupt:
    pass
  httpd.server_close()
  logging.info('Stopping httpd...\n')


if __name__ == '__main__':
  from sys import argv

  if len(argv) == 2:
    run(port=int(argv[1]))
  else:
    run()

```

tehnerd replied:
for now i will just move back all TE related logic. it was not part of original diff anyway, so if needed - this could be handled in separate, TE related diff.

tehnerd commented on test/extensions/filters/http/common/compressor/compressor_filter_test.cc:
```diff
@@ -584,34 +662,6 @@ TEST_P(IsMinimumContentLengthTest, Validate) {
-                    std::make_tuple("transfer-encoding", "test\t, chunked\t", false),
-                    std::make_tuple("x-garbage", "no_value", true)));
-
-TEST_P(IsTransferEncodingAllowedTest, Validate) {
```
will move back

tehnerd replied:
wait. no. this is transfer-encoding tests. already encoded content would have "content-encoding" headers. not transfer-encoding. this is why i've removed it

rojkov replied:
The upstream can set either "content-encoding" (encoding as a property of the resource) or "transfer-encoding" (encoding as a property of the message). In both cases the payload would be compressed, but in the latter case "content-encoding" would be missing.

alyssawilk commented on source/extensions/filters/http/common/compressor/compressor.cc:
```diff
@@ -156,7 +157,14 @@ Http::FilterHeadersStatus CompressorFilter::decodeHeaders(Http::RequestHeaderMap
-  if (!end_stream && request_config.compressionEnabled() &&
+  // temp variable to preserve old behavior w/ http upgrade. should be removed when
+  // enable_compression_without_content_length_header runtime variable would be removed
+  const bool allowUpgrade =
```
-> is_upgrade_allowed here and below?

alyssawilk commented on source/extensions/filters/http/common/compressor/compressor.cc:
```diff
@@ -506,9 +521,12 @@ bool CompressorFilterConfig::DirectionConfig::isMinimumContentLength(
-                                   Http::Headers::get().TransferEncodingValues.Chunked);
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.enable_compression_without_content_length_header")) {
+    // returning true to account for HTTP/2 where content-length is optional
```
and HTTP/1.1 chunked bodies, no?

alyssawilk replied:
was this one overlooked?

tehnerd replied:
could you please elaborate on the last comment?. w/ this new diff we would compress even if there is no "transfer-encoding=chunked" header (if runtime is set to true) (aka there is no point to do separate check for chunked. because we would have logic like:
```
if true || chunked {
  return true;
}
```
this check would always be true
. if runtime set to false we would fall back to previous behavior  aka "return (chunked in headers?)"

alyssawilk replied:
Yes, true.  Sorry, I meant to make the point that the comment implied that the return only applies to HTTP/2 but it applies to HTTP/1.1 without content length too.  How about something like

Return true to ignore the minimum length configuration if no content length is specified in the headers?

alyssawilk commented on test/extensions/filters/http/common/compressor/compressor_filter_test.cc:
```diff
@@ -233,6 +238,82 @@ TEST_F(CompressorFilterTest, CompressRequest) {
+)EOF");
+  doRequestCompression({{":method", "post"}}, false);
+  Http::TestResponseHeaderMapImpl headers{{":status", "200"}};
+  doResponseNoCompression(headers);
```
your call but I'd be inclined to combine the request and response to 
CompressNoContentLength

alyssawilk commented on test/extensions/filters/http/common/compressor/websocket_with_compressor_integration_test.cc:
```diff
@@ -0,0 +1,248 @@
+              *bootstrap.mutable_static_resources()->mutable_clusters(0), protocol_options);
+        });
+  }
+  if (downstreamProtocol() != Http::CodecClient::Type::HTTP1) {
```
why downstream protocol checks here?

alyssawilk commented on test/extensions/filters/http/common/compressor/websocket_with_compressor_integration_test.cc:
```diff
@@ -0,0 +1,248 @@
+  codec_client_->close();
+}
+
+TEST_P(WebsocketWithCompressorIntegrationTest, RouteSpecificUpgrade) {
```
I think we can nix this test.

alyssawilk commented on test/extensions/filters/http/common/compressor/websocket_with_compressor_integration_test.cc:
```diff
@@ -0,0 +1,248 @@
+
+// Technically not a websocket tests, but verifies normal upgrades have parity
+// with websocket upgrades
+TEST_P(WebsocketWithCompressorIntegrationTest, NonWebsocketUpgrade) {
```
can we also snag

 ProxyingConnectIntegrationTest, ProxyConnect

to make sure CONNECT requests and responses are handled correctly too?

alyssawilk reviewed (commented):
Cool, looking good!

tehnerd commented:
have not added itests for CONNECT as requested by @alyssawilk. so probably no need to review right now

tehnerd commented:
ok. i've added analog of (ProxyingConnectIntegrationTest, ProxyConnect) (the difference is that in initialize compression filter was also added). i think i've addressed all comments and this is ready for review

tehnerd commented:
looking on "verify examples" failed check. looks like it is related to test environment:
```
|[0m npm WARN tar ENOSPC: no space left on device, write
2021-01-27T22:50:17.2537235Z [36mwasm_compile_update_1  |[0m npm WARN tar ENOSPC: no space left on device, write
2021-01-27T22:50:17.2538088Z [36mwasm_compile_update_1  |[0m npm WARN tar ENOSPC: no space left on device, write
2021-01-27T22:50:17.2538826Z [36mwasm_compile_update_1  |[0m npm ERR! code ENOSPC
2021-01-27T22:50:17.2539500Z [36mwasm_compile_update_1  |[0m npm ERR! syscall write
2021-01-27T22:50:17.2540165Z [36mwasm_compile_update_1  |[0m npm ERR! errno -28
2021-01-27T22:50:17.2540938Z [36mwasm_compile_update_1  |[0m npm ERR! nospc ENOSPC: no space left on device, write
2021-01-27T22:50:17.2541841Z [36mwasm_compile_update_1  |[0m npm ERR! nospc There appears to be insufficient space on your system to finish.
2021-01-27T22:50:17.2542697Z [36mwasm_compile_update_1  |[0m npm ERR! nospc Clear up some disk space and try again.
```

alyssawilk commented on source/extensions/filters/http/common/compressor/compressor.cc:
```diff
@@ -156,7 +157,14 @@ Http::FilterHeadersStatus CompressorFilter::decodeHeaders(Http::RequestHeaderMap
+  // temp variable to preserve old behavior w/ http upgrade. should be removed when
+  // enable_compression_without_content_length_header runtime variable would be removed
+  const bool is_upgrade_allowed =
+      (!Http::Utility::isUpgrade(headers) && !Http::Utility::isH2UpgradeRequest(headers)) ||
```
hm, I'd think this should be isUpgradeRequest, and I'm wondering why the tests are working correctly as is.  Can you take a look and make sure we get the right function here and have regression tests that fail with things as they are?

tehnerd replied:
agree. will rework naming. should be "is_not_upgrade" . but yes. the logic is correct. w/o this line integration tests for upgrade are failing with:
```
Expected: header map equal ignore order 0x7ffde7a7cf48
  Actual: 0x7ffde7a7ce78 (of type Envoy::Http::TestResponseHeaderMapImpl*), 
==============================Expected header map:==============================
':status', '101'
'connection', 'upgrade'
'upgrade', 'foo'
=======================is not equal to actual header map:=======================
':status', '101'
'vary', 'Accept-Encoding'
'upgrade', 'foo'
'connection', 'upgrade'
================================================================================

Stack trace:
```

aka envoy is adding "vary" header to signal that potentially (if accept-encoding was set) it could compress such request/response. (and from my understanding of initial reported issue  -that was the actuall issue. that browser was not expecting this header. nothing was actually compressed (as i cannot see from code how that could happen if accept-encoding was not set in request)).

alyssawilk replied:
The HTTP/2 codec is supposed to transform HTTP/2 style upgrades to HTTP/1 style for the lifetime of the request in Envoy, so it should be impossible to look like an H2 upgrade request at this point in the pipeline.  Do you have an integration test where isH2UpgradeRequest is true?  If so I'll download the patch and sort out why my assumptions are off :-P

tehnerd replied:
nope. was just saw it in utility and just added as well. was not aware that it will make h2 looks like h1 in codec. will remove

alyssawilk commented on source/extensions/filters/http/compressor/BUILD:
```diff
@@ -27,6 +27,7 @@ envoy_cc_extension(
     name = "config",
     srcs = ["config.cc"],
     hdrs = ["config.h"],
+    extra_visibility = ["//test/extensions:__subpackages__"],
```
was this still necessary, and if so can we narrow it as much as possible?

tehnerd replied:
will remove. i think i missed this one during previous cleanup. thanks for catching

alyssawilk reviewed (commented):
Thanks for all the progress here!   One high level comment about the upgrade checks - I'll take a look at the tests once we have that sorted out.

tehnerd commented:
re on :envoy-presubmit. it is faling on deps validaiton w/
```
Evaluation of query "deps(//test/...)" failed: errors were encountered while computing transitive closure
```

no idea what that means, but would greatly appreciate if someone point me out what i've broke there (if any);)

tehnerd commented:
ok. looks like it (bazel deps(//test/...) is broken on "main" as well and i just picked it up during fetch/merge

alyssawilk commented on source/extensions/filters/http/common/compressor/compressor.cc:
```diff
@@ -215,13 +223,19 @@ void CompressorFilter::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallba
   }
 }
-
 Http::FilterHeadersStatus CompressorFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
```
super nit, it'd be nice to reinstate the whitespace here

tehnerd replied:
sure. surprised that fix_format didn't pick it up

alyssawilk commented on source/extensions/filters/http/common/compressor/compressor.cc:
```diff
@@ -506,7 +520,11 @@ bool CompressorFilterConfig::DirectionConfig::isMinimumContentLength(
-
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.enable_compression_without_content_length_header")) {
+    // return true to ignore the minimum length configuration if no content-length header presents
```
header presents -> header is present.

alyssawilk commented on test/extensions/filters/http/common/compressor/compressor_integration_tests.cc:
```diff
@@ -0,0 +1,296 @@
+
+  // Make sure that even once the response has started, that data can continue to go upstream.
+  codec_client_->sendData(*request_encoder_, "hello", false);
+  ASSERT_TRUE(upstream_request_->waitForData(*dispatcher_, 5));
```
Unlikely as it is that the compressed data would be the same size can we check for payload just in case?

ASSERT_TRUE(upstream_request_->waitForData(*dispatcher_, "hello"));

alyssawilk commented on test/extensions/filters/http/common/compressor/compressor_integration_tests.cc:
```diff
@@ -0,0 +1,296 @@
+
+  // Also test upstream to downstream data.
+  upstream_request_->encodeData(12, false);
+  response_->waitForBodyData(12);
```
upstream_request_->encodeData("world", false);
EXPECT_EQ("world", response_->body());

alyssawilk reviewed (commented):
LGTM modulo a few nits, and needs a mainmerge.  Snow would you be  up for final pass once those are done?

tehnerd commented:
hmm. merge picked up something which was deleted from current.rst. going to fix this and push again

alyssawilk reviewed (dismissed):
LGTM!  Thanks for all of your work on this!

tehnerd commented:
no changes in the last push. just merging upstream.

tehnerd commented:
meh. picked up removed change. going repush in a few

tehnerd commented:
actually seems like an issue w/ CI pipeline. envoy-presubmit failing with:
```
Generating script.
Script contents:
ci/run_envoy_docker.sh 'ci/check_and_fix_format.sh'
========================== Starting Command Output ===========================
/bin/bash --noprofile --norc /home/vsts/work/_temp/8870ce44-22a0-4574-95e8-b48de72c5192.sh
Unable to find image 'envoyproxy/envoy-build-ubuntu:c8fa4235714003ba0896287ee2f91cae06e0e407' locally
docker: Error response from daemon: unauthorized: authentication required.
See 'docker run --help'.
##[error]Bash exited with code '125'.
```

rojkov reviewed (dismissed):
Thank you! LGTM

rojkov commented:
/retest

tehnerd commented:
hmm. presubmit failing with:
```
TESTS FAILED:
run_examples@ ci/verify_examples.sh :35 (./websocket)
  > main@ ci/verify_examples.sh :40
```

however my diff does not touch anything websocket related (core part of it. only handling upgrade in compressor, but looks like this filter does not even enabled in examples:

```
tehnerd@amanda:~/projects/envoy$ grep -R compres ./examples/websocket/
tehnerd@amanda:~/projects/envoy$ grep -R compres ./examples/
tehnerd@amanda:~/projects/envoy$ grep -R gzip ./examples/
tehnerd@amanda:~/projects/envoy$ grep -R compres ./examples/
```

so no idea why it is failing.

rojkov commented:
/retest

tehnerd commented:
@rojkov  @alyssawilk  does it require anything else from my side? or we are just waiting for some one to merge this diff in?

tehnerd commented:
oh. looks like upstream changed and new merge required.

tehnerd commented:
presubmit failed with

alyssawilk commented:
I signed off but for most PRs we have 2 maintainers review.  I've tagged @snowp who was out late last week but should be able to do a pass this week.

tehnerd commented:
Thanks for the explanation, @alyssawilk. Question: if GitHub marking pr as "merge conflict" - should I update it before second maintainer will look?(asking because it looks like that GitHub removes "reviewed" tag of the first person if PR is updated)

alyssawilk commented:
It's always worth resolving merge conflicts - even if it removes the LGTM stamp, generally we operate on a maintainer signing off, not the actual stamp.  If there's significant changes between first and second reviewer signing off, sometimes reviewer#2 will tag reviewer#1 towards the end but it's not the norm.

alyssawilk commented:
try one more merge to see if it fixes coverage, and ping for @snowp

snowp commented on source/extensions/filters/http/common/compressor/compressor.cc:
```diff
@@ -156,7 +157,12 @@ Http::FilterHeadersStatus CompressorFilter::decodeHeaders(Http::RequestHeaderMap
 
   const auto& request_config = config_->requestDirectionConfig();
-  if (!end_stream && request_config.compressionEnabled() &&
+  const bool is_not_upgrade =
+      !Http::Utility::isUpgrade(headers) ||
+      !Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.enable_compression_without_content_length_header");
```
Seems odd to have the value of `is_not_upgrade` be dependent on the runtime flag? Can we rename the variable to make a bit more sense?

tehnerd replied:
idea was that we would drop runtime flag eventually and wont need to rename. if you feel strong about it - i will try to rework this part

snowp commented on source/extensions/filters/http/common/compressor/compressor.cc:
```diff
@@ -221,7 +227,12 @@ Http::FilterHeadersStatus CompressorFilter::encodeHeaders(Http::ResponseHeaderMa
   const bool isEnabledAndContentLengthBigEnough =
       config.compressionEnabled() && config.isMinimumContentLength(headers);
-  const bool isCompressible = isEnabledAndContentLengthBigEnough &&
+  const bool is_not_upgrade =
+      !Http::Utility::isUpgrade(headers) ||
+      !Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.enable_compression_without_content_length_header");
```
Same here. Alternatively split it into two different variables

snowp commented on test/extensions/filters/http/common/compressor/compressor_integration_tests.cc:
```diff
@@ -0,0 +1,297 @@
+#include "test/extensions/filters/http/common/compressor/compressor_integration_tests.h"
```
I would love a high level explanation of what we're testing in this file. I think it's that upgrades work through the compressor filter, but its not clear without reading a lot of code what kind of interactions (if any?) we're trying to validate

tehnerd replied:
ok. will add

snowp reviewed (changes requested):
Thanks this seems reasonable, just a few comments

snowp commented on test/extensions/filters/http/common/compressor/compressor_integration_tests.h:
```diff
@@ -8,6 +8,10 @@
 namespace Envoy {
 
+/*
+ Integration test to make sure that new logic, which does not care about content-length header
```
This comment reads like a PR comment, not a code comment. Can we phrase this in a way that will age better?

For example, `Test verifying that we don't add any additional headers when the upstream doesn't respond with a content-length header` or something to that effect

snowp commented on test/extensions/filters/http/common/compressor/compressor_integration_tests.h:
```diff
@@ -34,6 +38,10 @@ class WebsocketWithCompressorIntegrationTest : public HttpProtocolIntegrationTes
 };
 
+/*
+ Integration test to make sure that new logic, which does not care about content-length header
```
Same here, the "new logic" is only new right now

snowp reviewed (changes requested):
Thanks for iterating on this! Seems like CI is broken

snowp reviewed (approved):
Thanks!
